File: cpp/100SameTree/100SameTree.cpp
#include <sstream>
#include <iostream>
#include <string>
#include <vector>
#include <stack>
#include <unordered_map>
#include <algorithm>
#include <limits>

using namespace std;

/*
Given two binary trees, write a function to check if they are equal or not.

Two binary trees are considered equal if they are structurally identical and the nodes have the same value.
*/

// Definition for binary tree
struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};

class Solution {
public:
    bool isSameTree(TreeNode *p, TreeNode *q) {
      if(p==nullptr&&q==nullptr) return true;
      if((p==nullptr&&q!=nullptr)||(p!=nullptr&&q==nullptr)) return false;
      return p->val==q->val && isSameTree(p->left, q->left)&&isSameTree(p->right, q->right);
    }
};


int main(){
	return 0;
}

File: src/100SameTree/100SameTree.rs
// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn is_same_tree(p: Option<Rc<RefCell<TreeNode>>>, q: Option<Rc<RefCell<TreeNode>>>) -> bool {
        match (p,q) {
            (Some(pt), Some(qt)) => {
                let mut ptn = pt.borrow_mut();
                let mut qtn = qt.borrow_mut();
                ptn.val == qtn.val && Self::is_same_tree(ptn.left.take(), qtn.left.take()) && Self::is_same_tree(ptn.right.take(), qtn.right.take())
            }
            (None, None) => true,
            _ => false,
        }
    }
}
